libkmod/libkmod-elf.c

```diff
@@ -248,6 +269,56 @@ static inline int elf_get_section_info(const struct kmod_elf *elf, uint16_t idx,
 	return -EINVAL;
 }
 
+static void kmod_elf_save_sections(struct kmod_elf *elf)
+{
+	const uint16_t all_sec = (1 << KMOD_ELF_SECTION_MAX) - 1;
+	uint16_t found_sec = 0;
+
+	for (uint16_t i = 1; i < elf->header.section.count && found_sec != all_sec; i++) {
+		uint64_t off, size;
+		const char *n;
+		int err = elf_get_section_info(elf, i, &off, &size, &n);
+		if (err < 0)
+			continue;
+		if (streq("__ksymtab_strings", n)) {
+			elf->sections[KMOD_ELF_SECTION_KSYMTAB].offset = off;
+			elf->sections[KMOD_ELF_SECTION_KSYMTAB].size = size;
+			found_sec |= 1 << KMOD_ELF_SECTION_KSYMTAB;
+			continue;
+		}
```

Comment:
These if-blocks can be combined into one, put into another for-loop. The name itself can be retrieved through loop-variable from `section_name_map`. Having just one if-block makes it easier to audit the code.

Reply:
This code behaves differently than current implementation and kernel code. Right now, kmod and kernel take the first section information if multiple entries with the same name exist, but this code would take the last. To avoid any form of confusion between kmod and kernel, let's use only the first entry and add a warning through ELF_DBG (if you want to).

Reply:
Manual creation of module files. You could set the name offset of an already existing section to the same offset of another one. Or just add a completely new section etc.

Definitely no valid files, but let's make sure that it's not possible to let kmod read other sections than the kernel would.